Supersede MR::printf()

In trying to build #3315 with sanitisers for testing:

```text
[488/589] Linking CXX executable bin/mrtrix-unit-tests
FAILED: bin/mrtrix-unit-tests testing/unit_tests/mrtrix-unit-tests[1]_tests.cmake /home/unimelb.edu.au/robertes/src/worktrees/data_to_image_buffer/build_msan/testing/unit_tests/mrtrix-unit-tests[1]_tests.cmake 
: && /usr/bin/clang++-21 -fsanitize=memory -g  testing/unit_tests/CMakeFiles/mrtrix-unit-tests.dir/erfinv_tests.cpp.o testing/unit_tests/CMakeFiles/mrtrix-unit-tests.dir/icls_tests.cpp.o testing/unit_tests/CMakeFiles/mrtrix-unit-tests.dir/ordered_queue_tests.cpp.o testing/unit_tests/CMakeFiles/mrtrix-unit-tests.dir/parse_axes_tests.cpp.o testing/unit_tests/CMakeFiles/mrtrix-unit-tests.dir/parse_ints_tests.cpp.o testing/unit_tests/CMakeFiles/mrtrix-unit-tests.dir/platform_tests.cpp.o testing/unit_tests/CMakeFiles/mrtrix-unit-tests.dir/roi_tests.cpp.o testing/unit_tests/CMakeFiles/mrtrix-unit-tests.dir/sh_precomputer_tests.cpp.o testing/unit_tests/CMakeFiles/mrtrix-unit-tests.dir/shuffle_tests.cpp.o testing/unit_tests/CMakeFiles/mrtrix-unit-tests.dir/to_tests.cpp.o testing/unit_tests/CMakeFiles/mrtrix-unit-tests.dir/gputests.cpp.o -o bin/mrtrix-unit-tests  -Wl,-rpath,/home/unimelb.edu.au/robertes/src/worktrees/data_to_image_buffer/build_msan/cpp/core:/home/unimelb.edu.au/robertes/src/worktrees/data_to_image_buffer/build_msan/_deps/slang-src/lib  cpp/core/libmrtrix-core.so  cpp/core/libmrtrix-executable-version.a  lib/libgtest_main.a  /usr/lib/x86_64-linux-gnu/libpng.so  /usr/lib/x86_64-linux-gnu/libz.so  /usr/lib/x86_64-linux-gnu/libfftw3.so  cpp/core/libmrtrix-version.a  _deps/dawn-src/lib/libwebgpu_dawn.a  -ldl  _deps/slang-src/lib/libslang-compiler.so.0.2026.5.1  lib/libgtest.a && cd /home/unimelb.edu.au/robertes/src/worktrees/data_to_image_buffer/build_msan/testing/unit_tests && /usr/bin/cmake -D TEST_TARGET=mrtrix-unit-tests -D TEST_EXECUTABLE=/home/unimelb.edu.au/robertes/src/worktrees/data_to_image_buffer/build_msan/bin/mrtrix-unit-tests -D TEST_EXECUTOR= -D TEST_WORKING_DIR=/home/unimelb.edu.au/robertes/src/worktrees/data_to_image_buffer/build_msan/testing/unit_tests -D TEST_EXTRA_ARGS= -D "TEST_PROPERTIES=LABELS;unittest" -D TEST_PREFIX= -D TEST_SUFFIX= -D TEST_FILTER= -D NO_PRETTY_TYPES=FALSE -D NO_PRETTY_VALUES=FALSE -D TEST_LIST=mrtrix-unit-tests_TESTS -D CTEST_FILE=/home/unimelb.edu.au/robertes/src/worktrees/data_to_image_buffer/build_msan/testing/unit_tests/mrtrix-unit-tests[1]_tests.cmake -D TEST_DISCOVERY_TIMEOUT=5 -D TEST_XML_OUTPUT_DIR= -P /usr/share/cmake-3.22/Modules/GoogleTestAddTests.cmake
==2892907==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x5555559ab2ba in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> MR::str<double>(double const&, int) /home/unimelb.edu.au/robertes/src/worktrees/data_to_image_buffer/cpp/core/mrtrix.h:126:5
    #1 0x7ffff2d643b1 in __cxx_global_var_init.4 /home/unimelb.edu.au/robertes/src/worktrees/data_to_image_buffer/cpp/core/dwi/fmls.cpp:27:34
    #2 0x7ffff2d65f07 in _GLOBAL__sub_I_fmls.cpp /home/unimelb.edu.au/robertes/src/worktrees/data_to_image_buffer/cpp/core/dwi/fmls.cpp
    #3 0x7ffff7fc947d in call_init elf/dl-init.c:70:3
    #4 0x7ffff7fc9567 in call_init elf/dl-init.c:33:6
    #5 0x7ffff7fc9567 in _dl_init elf/dl-init.c:117:5
    #6 0x7ffff7fe32c9  (/lib64/ld-linux-x86-64.so.2+0x202c9) (BuildId: 8cfa19934886748ff4603da8aa8fdb0c2402b8cf)

SUMMARY: MemorySanitizer: use-of-uninitialized-value /home/unimelb.edu.au/robertes/src/worktrees/data_to_image_buffer/cpp/core/mrtrix.h:126:5 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> MR::str<double>(double const&, int)
Exiting
CMake Error at /usr/share/cmake-3.22/Modules/GoogleTestAddTests.cmake:83 (message):
  Error running test executable.

    Path: '/home/unimelb.edu.au/robertes/src/worktrees/data_to_image_buffer/build_msan/bin/mrtrix-unit-tests'
    Result: 1
    Output:
      

Call Stack (most recent call first):
  /usr/share/cmake-3.22/Modules/GoogleTestAddTests.cmake:179 (gtest_discover_tests_impl)
```

This looks to me like `MR::printf()` (`mrtrix.h:126`) is using a kind of C equivalent of variadic templates?
Either this should be replaced with a C++ variadic template version, or the whole thing should be usurped using eg. #3125.

--- a/cpp/core/mrtrix.cpp
+++ b/cpp/core/mrtrix.cpp
@@ -16,7 +16,6 @@
 
 #include "mrtrix.h"
 
-#include <cstdarg>
 #include <string_view>
 
 namespace MR {
@@ -165,19 +164,6 @@ std::string uppercase(std::string_view string) {
   ret.resize(string.size());
   transform(string.begin(), string.end(), ret.begin(), toupper);
   return ret;
-}
-
-std::string printf(const char *format, ...) { // check_syntax off
-  size_t len = 0;
-  va_list list1, list2;
-  va_start(list1, format);
-  va_copy(list2, list1);
-  len = vsnprintf(nullptr, 0, format, list1) + 1;
-  va_end(list1);
-  VLA(buf, char, len);
-  vsnprintf(buf, len, format, list2);
-  va_end(list2);
-  return buf;
 }
 
 std::string strip(std::string_view string, std::string_view ws, bool left, bool right) {

--- a/cpp/core/mrtrix.h
+++ b/cpp/core/mrtrix.h
@@ -78,7 +78,15 @@ std::string lowercase(std::string_view string);
 //! return uppercase version of string
 std::string uppercase(std::string_view string);
 
-std::string printf(const char *format, ...); // check_syntax off
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-security"
+template <typename... Args> std::string printf(const std::string format, Args... args) {
+  const int len = std::snprintf(nullptr, 0, format.c_str(), args...) + 1;
+  VLA(buf, char, len);
+  std::snprintf(buf, len, format.c_str(), args...);
+  return std::string(buf);
+}
+#pragma GCC diagnostic pop
 
 std::string strip(std::string_view string, std::string_view ws = {" \0\t\r\n", 5}, bool left = true, bool right = true);
 

